server/src/main/java/org/elasticsearch/index/IndexVersion.java

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index;
+
+import org.apache.lucene.util.Version;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.core.Assertions;
+
+import java.lang.reflect.Field;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * The index version.
+ * <p>
+ * Prior to 8.8.0, the node {@link Version} was used everywhere. This class separates the index format version
+ * from the running node version.
+ * <p>
+ * Each index version constant has an id number, which for versions prior to 8.9.0 is the same as the release version
+ * for backwards compatibility. In 8.9.0 this is changed to an incrementing number, disconnected from the release version.
```

Comment:
While that is the intention, I think it depends on if this can be decoupled before 8.9.0 is branched. I'm not sure what to change here, but if this changes, we need to come back and update the comment.

Reply:
Of course - we had the same issue in `TransportVersion` too